Add _valtype support for RadialDecay with HyperDual input

DexuanZhou opened the pull request:
This PR adds _valtype of RadialDecay to support HyperDual{T} input types from [HyperDualNumbers.jl](https://github.com/JuliaDiff/HyperDualNumbers.jl).

This is required for computing second-order derivatives (e.g., Laplacians) of wavefunctions using HyperDual-based automatic differentiation.

cortner reviewed (changes requested):
Should these three definitions of `_valtype` be merged into a single one? 
```julia
_valtype(::RadialDecay, T::Type{<: Number}, args...) = T
```

Can you try this?

cortner commented:
@DexuanZhou  - can you look at this briefly? If what I suggest works, then please update the PR and I'll merge it and then tag together with the bugfixes I'm currently working on.

cortner commented:
coming back to this. Sorry to be a pain, but I don't want to merge it like this yet. I know it looks like a trivial change but many changes like this will lead to proliferation of hacks. 

(1) please add a test that fails without this addition. 
(2) test what is needed to make the tests pass
(3) test in particular also the calling convention with ps and st, not just the straight evaluate.

cortner commented:
@DexuanZhou  --  what's the status of this PR?

DexuanZhou commented:
> Should these three definitions of `_valtype` be merged into a single one?
> 
> ```julia
> _valtype(::RadialDecay, T::Type{<: Number}, args...) = T
> ```
> 
> Can you try this?

Yes, those three _valtype methods can be merged into a single definition without breaking anything.

DexuanZhou commented:
I added tests for evaluate methods that support (basis, x, ps, st) inputs, unified three valtype methods into one, but kept the following two RadialDecay valtype methods separate:
```julia
_valtype(::RadialDecay, T::Type{<: Number}, args...) = T

_valtype(::RadialDecay, T::Type{<: Real}, ps, st) =
    promote_type(T, eltype(ps.ζ), eltype(ps.D))
```
These cannot be merged because doing so breaks one of the Lux tests.
And I added HyperDual tests to check consistency between HyperDual differentiation and the analytical derivatives (evaluate_ed), for Gaussian, Slater, and STO bases.

cortner commented:
Thanks, I'll take a look later. I also wonder whether we should make this an extension rather than keep it in the core package.

DexuanZhou commented:
I think making it an extension makes sense.

cortner commented:
If I see it correctly you've now introduced HyperDualNumbers _ONLY_ in the tests. I think that doesn't require an extension but it suggests moving `test_hyperdual_consistency` and the accompanying functions into a test script, e.g. `hyperdual_testutils.jl` that can then be loaded via `include`. Then hyperduals can be part of the test dependencies but be entirely removed from the package dependencies and we don't have to mess around with extensions either
